core/src/main/scala/kafka/log/Log.scala

```diff
@@ -744,9 +744,9 @@ class Log(@volatile var dir: File,
   private def updateLogEndOffset(messageOffset: Long): Unit = {
     nextOffsetMetadata = LogOffsetMetadata(messageOffset, activeSegment.baseOffset, activeSegment.size)
 
-    // Update the high watermark in case it has gotten ahead of the log end offset
-    // following a truncation.
-    if (highWatermark > messageOffset) {
+    // Update the high watermark in case it has gotten ahead of the log end offset following a truncation
+    // or if a new segment has been rolled and the offset metadata needs to be updated.
+    if (highWatermark >= messageOffset) {
```

Comment:
Would it make sense to make this more explicit by the caller? It's clear that a hw that is ahead of the offset needs updating, but less so for the case where it's the same. It seems like we are relying on implicit implementation details that make the code harder to reason about.

Would it help if the caller (there are only 4) specified if the log was rolled and/or truncated? Or is there risk of this condition happening independently of the caller?

Reply:
It's a fair point. I was shooting for a minimal diff here to make merging easier, but there's room for improvement in making expectations more explicit. I'll think about this and submit a follow-up if I can come up with a nicer way to handle this.